static/app/utils/useUserTeams.tsx

```diff
@@ -0,0 +1,65 @@
+import {useEffect} from 'react';
+import uniqBy from 'lodash/uniqBy';
+
+import OrganizationStore from 'sentry/stores/organizationStore';
+import TeamStore from 'sentry/stores/teamStore';
+import {useLegacyStore} from 'sentry/stores/useLegacyStore';
+import type {Team} from 'sentry/types';
+import {isActiveSuperuser} from 'sentry/utils/isActiveSuperuser';
+import {ApiQueryKey, useApiQuery} from 'sentry/utils/queryClient';
+
+interface UseTeamsResult {
+  isError: boolean | null;
+  isLoading: boolean;
+  teams: Team[];
+}
+
+function buildUserTeamsQueryKey(orgSlug: string): ApiQueryKey {
+  return [`/organizations/${orgSlug}/user-teams/`];
+}
+
+/**
+ * Fetches the user's teams if they aren't already in the TeamStore.
+ * @returns an array of user's teams, isLoading, and isError
+ *
+ * example usage:
+ * ```ts
+ * const {teams, isLoading, isError} = useUserTeams();
+ * ```
+ */
+export function useUserTeams(): UseTeamsResult {
+  const {organization} = useLegacyStore(OrganizationStore);
+  const storeState = useLegacyStore(TeamStore);
+
+  // Wait until the store has loaded to start fetching
+  const shouldConsiderFetching = !!organization?.slug && !storeState.loading;
+  // Only fetch if there are missing teams
+  const hasMissing = !storeState.loadedUserTeams;
+  const queryEnabled = shouldConsiderFetching && hasMissing;
+
+  const {
+    data: additionalTeams = [],
+    isLoading,
+    isError,
+  } = useApiQuery<Team[]>(buildUserTeamsQueryKey(organization?.slug ?? ''), {
+    staleTime: 0,
+    enabled: queryEnabled,
+  });
+
+  // Save fetched teams to the team store
+  useEffect(() => {
+    if (additionalTeams.length > 0) {
+      TeamStore.loadUserTeams(additionalTeams);
+    }
+  }, [additionalTeams]);
+
+  const isSuperuser = isActiveSuperuser();
+  const allTeams = uniqBy([...storeState.teams, ...additionalTeams], team => team.id);
+  const teams = isSuperuser ? allTeams : allTeams.filter(t => t.isMember);
```

Comment:
since this is returned, we should probably make it stable. As is, we'll be recreating the array each time and might end up triggering effect/react updates for components that use this hook unnecessarily